### PR TITLE
Add support for UC volumes as artifact repository 

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/12545?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12545/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12545
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Taking over #12072. Adding support for UC volumes as artifact repository .

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

```python
import contextlib
import tempfile
from pathlib import Path
from unittest import mock

import numpy as np
import requests
from sklearn.linear_model import LogisticRegression

import mlflow
from mlflow.models import infer_signature

original_session_request = requests.Session.request


def p(*args, **kwargs):
    res = original_session_request(*args, **kwargs)
    print("-" * 30)
    print(args[1:], kwargs, res)
    return res


@contextlib.contextmanager
def spy_request():
    with mock.patch("requests.Session.request", p):
        yield


def main():
    EXP_NAME = "/Users/harutaka.kawamura@databricks.com/uc-volumes-artifact"
    CATALOG = "ml"
    SCHEMA = "haru"
    VOLUME = "uc-volumes-artifact"
    ARTIFACT_PATH = f"dbfs:/Volumes/{CATALOG}/{SCHEMA}/{VOLUME}"

    mlflow.set_tracking_uri("databricks")
    mlflow.set_registry_uri("databricks-uc")

    if mlflow.get_experiment_by_name(EXP_NAME) is None:
        mlflow.create_experiment(name=EXP_NAME, artifact_location=ARTIFACT_PATH)
    mlflow.set_experiment(EXP_NAME)

    X = np.array([-2, -1, 0, 1, 2, 1]).reshape(-1, 1)
    y = np.array([0, 0, 1, 1, 1, 0])
    lr = LogisticRegression().fit(X, y)

    predictions = lr.predict(X)
    signature = infer_signature(X, predictions)

    with mlflow.start_run() as run:
        with spy_request():
            model_info = mlflow.sklearn.log_model(
                lr,
                "model",
                signature=signature,
                registered_model_name=f"{CATALOG}.{SCHEMA}.uc-volumes-artifact",
            )

            client = mlflow.MlflowClient()
            print(client.list_artifacts(run_id=model_info.run_id))
            print(client.list_artifacts(run_id=model_info.run_id, path="model"))
            print(client.list_artifacts(run_id=model_info.run_id, path="model/MLmodel"))

            with tempfile.TemporaryDirectory() as tmpdir:
                mlflow.artifacts.download_artifacts(run.info.artifact_uri, dst_path=tmpdir)
                for p in Path(tmpdir).rglob("*"):
                    print(p)


if __name__ == "__main__":
    main()

```

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/store/artifact/uc_volume_artifact_repo.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+            endpoint=f"{DIR_API_ENDPOINT}{volume_path}", method="HEAD"
+        )
+        try:
+            return response.status_code == 200
+        except KeyError:
+            raise MlflowException(f"Volume path {volume_path} does not exist")
+
+    def _get_volumes_path(self, artifact_path):
+        return "/{}/{}".format(
+            strip_prefix(self.artifact_uri, "dbfs:/"),
```
Comment (WeichenXu123) [inline mlflow/store/artifact/uc_volume_artifact_repo.py:94]: why not using `urllib.parse.urlparse` ?

Comment (WeichenXu123) [reply mlflow/store/artifact/uc_volume_artifact_repo.py:94]: question: is the kind of url legal ? "dbfs:///xx/yy"

File: mlflow/store/artifact/uc_volume_artifact_repo.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            hostcreds_from_uri = get_databricks_host_creds(databricks_profile_uri)
+            self.get_host_creds = lambda: hostcreds_from_uri
+        else:
+            self.get_host_creds = _get_host_creds_from_default_store()
+
+    def _databricks_api_request(self, endpoint, method, **kwargs):
+        host_creds = self.get_host_creds()
+        return http_request_safe(host_creds=host_creds, endpoint=endpoint, method=method, **kwargs)
+
+    def _volumes_list_api(self, path):
```
Comment (WeichenXu123) [inline mlflow/store/artifact/uc_volume_artifact_repo.py:62]: The method name is confusing, it sends request to get listing result, so better name would be `_volumes_list`

File: mlflow/store/artifact/uc_volume_artifact_repo.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        host_creds = self.get_host_creds()
+        return http_request(
+            host_creds=host_creds, endpoint=f"{DIR_API_ENDPOINT}{path}", method="GET"
+        )
+
+    def _volumes_download(self, output_path, endpoint):
```
Comment (WeichenXu123) [inline mlflow/store/artifact/uc_volume_artifact_repo.py:68]: Do we want to support parallel downloading ?

File: mlflow/store/artifact/uc_volume_artifact_repo.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+            self._databricks_api_request(
+                endpoint=http_endpoint, method="PUT", data="", allow_redirects=False
+            )
+        else:
+            with open(local_file, "rb") as f:
+                self._databricks_api_request(
+                    endpoint=http_endpoint, method="PUT", data=f, allow_redirects=False
+                )
+
+    def log_artifacts(self, local_dir, artifact_path=None):
```
Comment (WeichenXu123) [inline mlflow/store/artifact/uc_volume_artifact_repo.py:120]: question about the retry behavior

Assuming log_artifact is interrupted (e.g. network issue), and part of files have been uploaded successfully, and user want to retry `log_artifact`, then for those files which have been uploaded to UC successfully, will the retry run of ` overwrite the files last time uploaded ? Or error occurs ?

**Comment by harupy** [comment]
@WeichenXu123 thanks for the review, but it's not ready for review yet. will ping you once it's ready!

File: mlflow/store/artifact/uc_volume_artifact_repo.py (hunk 1 above)
Comment (harupy) [reply mlflow/store/artifact/uc_volume_artifact_repo.py]: Good call. Filed #12580

File: mlflow/store/artifact/uc_volume_artifact_repo.py (hunk 2 above)
Comment (harupy) [reply mlflow/store/artifact/uc_volume_artifact_repo.py]: I think it comes from `mlflow/store/artifact/dbfs_artifact_repo.py`.

File: mlflow/store/artifact/uc_volume_artifact_repo.py (hunk 3 above)
Comment (harupy) [reply mlflow/store/artifact/uc_volume_artifact_repo.py]: That's a follow-up.

File: mlflow/store/artifact/uc_volume_artifact_repo.py (hunk 4 above)
Comment (harupy) [reply mlflow/store/artifact/uc_volume_artifact_repo.py]: overwrite

File: mlflow/store/artifact/uc_volumes_artifact_repo.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        host_creds = self.get_host_creds()
+        return http_request_safe(host_creds=host_creds, endpoint=endpoint, method=method, **kwargs)
+
+    def _volumes_list_api(self, path):
+        host_creds = self.get_host_creds()
+        return http_request(
+            host_creds=host_creds, endpoint=f"{DIR_API_ENDPOINT}{path}", method="GET"
+        )
+
+    def _volumes_download(self, output_path, endpoint):
```
Comment (BenWilson2) [inline mlflow/store/artifact/uc_volumes_artifact_repo.py:68]: can this be changed to a verb-noun pattern in keeping with method naming conventions?

File: mlflow/store/artifact/uc_volumes_artifact_repo.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+            finally:
+                response.close()
+
+    def _is_directory(self, artifact_path):
+        volume_path = (
+            self._get_volumes_path(artifact_path) if artifact_path else self._get_volumes_path("")
+        )
+        return self._volumes_is_dir(volume_path)
+
+    def _volumes_is_dir(self, volume_path):
```
Comment (BenWilson2) [inline mlflow/store/artifact/uc_volumes_artifact_repo.py:83]: verb-noun here as well. `_is_volumes_directory`

File: mlflow/store/artifact/uc_volumes_artifact_repo.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        response = self._databricks_api_request(
+            endpoint=f"{DIR_API_ENDPOINT}{volume_path}", method="HEAD"
+        )
+        try:
+            return response.status_code == 200
+        except KeyError:
```
Comment (BenWilson2) [inline mlflow/store/artifact/uc_volumes_artifact_repo.py:89]: Can we raise for status instead of catching a failure to access the status_code? Are we sure that the Volumes API doesn't return a status code if the volume doesn't exist? This feels a bit fragile. Surely we get a 5xx or a 4xx error, right?

File: mlflow/store/artifact/uc_volumes_artifact_repo.py [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+            raise MlflowException(f"Volume path {volume_path} does not exist")
+
+    def _get_volumes_path(self, artifact_path):
+        return "/{}/{}".format(
```
Comment (BenWilson2) [inline mlflow/store/artifact/uc_volumes_artifact_repo.py:93]: We probably want to use pathlib here instead of this interpolation handling

File: mlflow/store/artifact/uc_volumes_artifact_repo.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+    if not isinstance(store, RestStore):
+        raise MlflowException(
+            "Failed to get credentials for DBFS; they are read from the "
+            + "Databricks CLI credentials or MLFLOW_TRACKING* environment "
+            + "variables."
+        )
+    return store.get_host_creds
+
+
+def uc_volumes_artifact_repo_factory(artifact_uri):
```
Comment (BenWilson2) [inline mlflow/store/artifact/uc_volumes_artifact_repo.py:180]: can we rename this func?

File: mlflow/utils/rest_utils.py [hunk 10]
```diff
@@ -240,7 +243,9 @@ def verify_rest_response(response, endpoint):
             "API request to endpoint was successful but the response body was not "
             "in a valid JSON format"
         )
-        raise MlflowException(f"{base_msg}. Response body: '{response.text}'")
+        # no longer raising an exception for a succesfful operation
```
Comment (BenWilson2) [inline mlflow/utils/rest_utils.py:246]: Why is this removed?

File: mlflow/store/artifact/uc_volumes_artifact_repo.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+    if not is_valid_dbfs_uri(artifact_uri):
+        raise MlflowException(
+            "DBFS URI must be of the form dbfs:/Volumes/<path> or "
+            + "dbfs://profile@databricks/Volumes/<path>, but received "
+            + artifact_uri
+        )
+
+    cleaned_artifact_uri = artifact_uri.rstrip("dbfs:")
+    db_profile_uri = get_databricks_profile_uri_from_artifact_uri(cleaned_artifact_uri)
+    if (
```
Comment (BenWilson2) [inline mlflow/store/artifact/uc_volumes_artifact_repo.py:202]: Can this logic be cleaned up / made into a utility validator?

File: mlflow/store/artifact/uc_volumes_artifact_repo.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+    def list_artifacts(self, path=None):
+        volumes_path = self._get_volumes_path(path) if path else self._get_volumes_path("")
+        response = self._volumes_list_api(volumes_path)
+        try:
+            json_response = json.loads(response.text)
+        except ValueError:
+            raise MlflowException(
+                f"API request to list files under Volumes path {volumes_path} failed with "
+                f"status code {response.status_code}. Response body: {response.text}"
+            )
```
Comment (BenWilson2) [inline mlflow/store/artifact/uc_volumes_artifact_repo.py:141]: Shouldn't this be handled in the REST handler?

**Comment by harupy** [comment]
@BenWilson2 thanks for the review, but it's not ready for review.

File: mlflow/store/artifact/uc_volumes_artifact_repo.py (hunk 9 above)
Comment (harupy) [reply mlflow/store/artifact/uc_volumes_artifact_repo.py:210]: can you suggest the name?

**Comment by harupy** [comment]
@WeichenXu123 @BenWilson2 ready for review now!

File: mlflow/store/artifact/uc_volume_artifact_repo.py [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+    strip_scheme,
+)
+
+# https://docs.databricks.com/api/workspace/files
+DIRECTORIES_API_ENDPOINT = "/api/2.0/fs/directories"
+FILES_API_ENDPOINT = "/api/2.0/fs/files"
+DOWNLOAD_CHUNK_SIZE = 1024
+
+
+class UCVolumeArtifactRepository(ArtifactRepository):
```
Comment (BenWilson2) [inline mlflow/store/artifact/uc_volume_artifact_repo.py:31]:
```suggestion
class UCVolumesArtifactRepository(ArtifactRepository):
```

File: mlflow/store/artifact/uc_volume_artifact_repo.py [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+class UCVolumesArtifactRepository(ArtifactRepository):
+    """
+    Stores artifacts on UC Volumes using the Files REST API.
+    """
+
+    def __init__(self, artifact_uri):
+        if not is_valid_uc_volumes_uri(artifact_uri):
+            raise MlflowException(
+                message=(
+                    f"UC volume URI must be of the form "
```
Comment (BenWilson2) [inline mlflow/store/artifact/uc_volume_artifact_repo.py:40]:
```suggestion
                    f"UC Volumes URI must be of the form "
```
Might be more recognizable by the product name :)

File: mlflow/store/artifact/uc_volume_artifact_repo.py [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+
+        Returns:
+            The response from the API.
+
+        See also:
+            https://docs.databricks.com/api/workspace/files/download
+        """
+        endpoint = f"{FILES_API_ENDPOINT}{file_path}"
+        with open(output_path, "wb") as f:
+            with self._api_request(endpoint=endpoint, method="GET", stream=True) as resp:
```
Comment (BenWilson2) [inline mlflow/store/artifact/uc_volume_artifact_repo.py:117]: Do we need to disallow redirects when downloading?

Comment (harupy) [reply mlflow/store/artifact/uc_volume_artifact_repo.py:117]: yea, thanks for the catch!

Comment (harupy) [reply mlflow/store/artifact/uc_volume_artifact_repo.py:119]: In `mlflow/store/artifact/dbfs_artifact_repo.py`, redirects are only disallowed for `log_artifact` 🤔

https://github.com/mlflow/mlflow/blob/1533351efc56c68c79a39ee215a2747eef900855/mlflow/store/artifact/dbfs_artifact_repo.py#L119

Comment (BenWilson2) [reply mlflow/store/artifact/uc_volume_artifact_repo.py:117]: I suppose that makes sense for a data intercept security risk, which would be a primary attack vector on a compromised machine. SGTM

File: tests/store/artifact/test_uc_volumes_artifact_repo.py [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+def artifact_repo(monkeypatch):
+    return get_artifact_repository("dbfs:/Volumes/catalog/schema/volume/run_id/artifacts")
+
+
+@pytest.mark.parametrize(
+    "artifact_uri",
+    [
+        "dbfs:/Volumes/catalog/schema/volume/path",
+        "dbfs:/volumes/catalog/schema/volume/path",
+        "dbfs:/Volume/catalog/schema/volume/path",
```
Comment (BenWilson2) [inline tests/store/artifact/test_uc_volumes_artifact_repo.py:38]: +1 to validating the typo-aware match logic in the UC implementation

**Review by BenWilson2 — APPROVED** [review]
Looks great! Thanks for getting this fully implemented @harupy !

File: tests/store/artifact/test_uc_volumes_artifact_repo.py (hunk 16 above)
Comment (harupy) [reply tests/store/artifact/test_uc_volumes_artifact_repo.py:38]: This is not a typo. `Volume`, `volume`, and `volumes` are all alias:

https://docs.databricks.com/en/volumes/index.html#reserved-paths-for-volumes
